zio-kafka/src/main/scala/zio/kafka/utils/ExtraZStreamOps.scala

```diff
@@ -0,0 +1,89 @@
+package zio.kafka.utils
+
+import zio._
+import zio.stream._
+
+import java.time.Instant
+import scala.math.Ordered.orderingToOrdered
+
+object ExtraZStreamOps {
+  private sealed trait State
+  private final case class WaitingForProducer(producerDone: Promise[Nothing, Unit]) extends State
+  private final case class WaitingForConsumer(deadline: Instant)                    extends State
+  private case object ConsumerTimedOut                                              extends State
+
+  implicit final class ZStreamOps[R, E, A](private val stream: ZStream[R, E, A]) extends AnyVal {
+
+    /**
+     * Fails the stream with given error if it is not consumed (pulled) from, for some duration.
+     *
+     * Will also run workflow `onTimeout` when the timeout occurs.
+     *
+     * Note: since the wrapped stream only starts when pulling starts, we never time out waiting for the first pull
+     *
+     * Also see [[zio.stream.ZStream.timeoutFail]] for failing the stream doesn't _produce_ a value.
+     */
+    // noinspection SimplifySleepInspection
+    def consumeTimeoutFail[E1 >: E](
+      e: => E1
+    )(after: Duration)(onTimeout: ZIO[R, Nothing, Any])(implicit trace: Trace): ZStream[R, E1, A] =
```

Comment:
Avoids allocating things if the timeout never happens

```suggestion
    )(after: Duration)(onTimeout: => ZIO[R, Nothing, Any])(implicit trace: Trace): ZStream[R, E1, A] =
```